File: RunningStats.cpp
#include "RunningStats.hpp"
#include <math.h>

// https://www.geeksforgeeks.org/confidence-interval/
// assumes standard distribution
// 90% 1.645
// 95% 1.960
// 99% 2.576
static _float_t z_values[] = {1.645, 1.960, 2.576};

RunningStats::RunningStats() { Clear(); }

void RunningStats::Clear() {
  n = 0;
  M1 = M2 = M3 = M4 = 0.0;
}

void RunningStats::Push(_float_t x) {
  _float_t delta, delta_n, delta_n2, term1;

  long long n1 = n;
  n++;
  delta = x - M1;
  delta_n = delta / n;
  delta_n2 = delta_n * delta_n;
  term1 = delta * delta_n * n1;
  M1 += delta_n;
  M4 += term1 * delta_n2 * (n * n - 3 * n + 3) + 6 * delta_n2 * M2 -
        4 * delta_n * M3;
  M3 += term1 * delta_n * (n - 2) - 3 * delta_n * M2;
  M2 += term1;
}

long long RunningStats::NumDataValues() const { return n; }

_float_t RunningStats::Mean() const { return M1; }

_float_t RunningStats::Variance() const { return M2 / (n - 1.0); }

_float_t RunningStats::StandardDeviation() const { return sqrt(Variance()); }

_float_t RunningStats::Skewness() const {
  return sqrt(_float_t(n)) * M3 / pow(M2, 1.5);
}

_float_t RunningStats::Kurtosis() const {
  return _float_t(n) * M4 / (M2 * M2) - 3.0;
}

_float_t RunningStats::ConfidenceInterval(ci_t ci) {
  // if (n < 300)
  //   return NAN;
  // if (ci < CI90 || ci > CI99)
  //   return NAN;
  return z_values[ci] * StandardDeviation() / sqrt((_float_t)n);
}

RunningStats operator+( RunningStats const &a,  RunningStats const &b) {
  RunningStats combined;

  combined.n = a.n + b.n;

  _float_t delta = b.M1 - a.M1;
  _float_t delta2 = delta * delta;
  _float_t delta3 = delta * delta2;
  _float_t delta4 = delta2 * delta2;

  combined.M1 = (a.n * a.M1 + b.n * b.M1) / combined.n;

  combined.M2 = a.M2 + b.M2 + delta2 * a.n * b.n / combined.n;

  combined.M3 = a.M3 + b.M3 +
                delta3 * a.n * b.n * (a.n - b.n) / (combined.n * combined.n);
  combined.M3 += 3.0 * delta * (a.n * b.M2 - b.n * a.M2) / combined.n;

  combined.M4 = a.M4 + b.M4 +
                delta4 * a.n * b.n * (a.n * a.n - a.n * b.n + b.n * b.n) /
                    (combined.n * combined.n * combined.n);
  combined.M4 += 6.0 * delta2 * (a.n * a.n * b.M2 + b.n * b.n * a.M2) /
                     (combined.n * combined.n) +
                 4.0 * delta * (a.n * b.M3 - b.n * a.M3) / combined.n;

  return combined;
}

RunningStats &RunningStats::operator+=(const RunningStats &rhs) {
  RunningStats combined = *this + rhs;
  *this = combined;
  return *this;
}

File: js/RunningStats.ts
export enum ConfidenceInterval {
    CI90 = 0,
    CI95 = 1,
    CI99 = 2
}

export class RunningStats {
    private n: number = 0;
    private M1: number = 0; // First moment (mean)
    private M2: number = 0; // Second moment (variance)
    private M3: number = 0; // Third moment (skewness)
    private M4: number = 0; // Fourth moment (kurtosis)

    // Z-values for confidence intervals
    private static readonly Z_VALUES = [1.645, 1.960, 2.576]; // 90%, 95%, 99%

    constructor() {
        this.clear();
    }

    /**
     * Reset all statistics to initial state
     */
    clear(): void {
        this.n = 0;
        this.M1 = 0.0;
        this.M2 = 0.0;
        this.M3 = 0.0;
        this.M4 = 0.0;
    }

    /**
     * Add a new value to the statistics
     * @param x The value to add
     */
    push(x: number): void {
        const n1 = this.n;
        this.n++;

        const delta = x - this.M1;
        const delta_n = delta / this.n;
        const delta_n2 = delta_n * delta_n;
        const term1 = delta * delta_n * n1;

        this.M1 += delta_n;
        this.M4 += term1 * delta_n2 * (this.n * this.n - 3 * this.n + 3) +
                   6 * delta_n2 * this.M2 - 4 * delta_n * this.M3;
        this.M3 += term1 * delta_n * (this.n - 2) - 3 * delta_n * this.M2;
        this.M2 += term1;
    }

    /**
     * Get the number of values added
     * @return Number of data points
     */
    numDataValues(): number {
        return this.n;
    }

    /**
     * Get the mean of the values
     * @return The mean
     */
    mean(): number {
        return this.M1;
    }

    /**
     * Get the sample variance (n-1 denominator)
     * @return The sample variance
     */
    variance(): number {
        return (this.n > 1) ? this.M2 / (this.n - 1.0) : 0.0;
    }

    /**
     * Get the population variance (n denominator)
     * @return The population variance
     */
    populationVariance(): number {
        return (this.n > 0) ? this.M2 / this.n : 0.0;
    }

    /**
     * Get the standard deviation
     * @return The standard deviation
     */
    standardDeviation(): number {
        return Math.sqrt(this.variance());
    }

    /**
     * Get the skewness
     * @return The skewness
     */
    skewness(): number {
        if (this.n === 0 || this.M2 === 0) return 0.0;
        return Math.sqrt(this.n) * this.M3 / Math.pow(this.M2, 1.5);
    }

    /**
     * Get the kurtosis
     * @return The kurtosis
     */
    kurtosis(): number {
        if (this.n === 0 || this.M2 === 0) return 0.0;
        return (this.n * this.M4) / (this.M2 * this.M2) - 3.0;
    }

    /**
     * Get the confidence interval
     * @param ci The confidence interval type
     * @return The confidence interval value
     */
    confidenceInterval(ci: ConfidenceInterval): number {
        if (this.n < 30) return NaN; // Need sufficient sample size
        if (ci < ConfidenceInterval.CI90 || ci > ConfidenceInterval.CI99) return NaN;

        return RunningStats.Z_VALUES[ci] * this.standardDeviation() / Math.sqrt(this.n);
    }

    /**
     * Combine two RunningStats instances
     * @param other The other RunningStats instance to combine with
     * @return A new combined RunningStats instance
     */
    combine(other: RunningStats): RunningStats {
        const combined = new RunningStats();

        combined.n = this.n + other.n;

        if (combined.n === 0) return combined;

        const delta = other.M1 - this.M1;
        const delta2 = delta * delta;
        const delta3 = delta * delta2;
        const delta4 = delta2 * delta2;

        combined.M1 = (this.n * this.M1 + other.n * other.M1) / combined.n;

        combined.M2 = this.M2 + other.M2 + delta2 * this.n * other.n / combined.n;

        combined.M3 = this.M3 + other.M3 +
                      delta3 * this.n * other.n * (this.n - other.n) / (combined.n * combined.n);
        combined.M3 += 3.0 * delta * (this.n * other.M2 - other.n * this.M2) / combined.n;

        combined.M4 = this.M4 + other.M4 +
                      delta4 * this.n * other.n * (this.n * this.n - this.n * other.n + other.n * other.n) /
                      (combined.n * combined.n * combined.n);
        combined.M4 += 6.0 * delta2 * (this.n * this.n * other.M2 + other.n * other.n * this.M2) /
                       (combined.n * combined.n) +
                       4.0 * delta * (this.n * other.M3 - other.n * this.M3) / combined.n;

        return combined;
    }

    /**
     * Add another RunningStats instance to this one
     * @param other The other RunningStats instance to add
     * @return This instance for chaining
     */
    add(other: RunningStats): RunningStats {
        const combined = this.combine(other);
        this.n = combined.n;
        this.M1 = combined.M1;
        this.M2 = combined.M2;
        this.M3 = combined.M3;
        this.M4 = combined.M4;
        return this;
    }

    /**
     * Create a copy of this RunningStats instance
     * @return A new RunningStats instance with the same values
     */
    clone(): RunningStats {
        const copy = new RunningStats();
        copy.n = this.n;
        copy.M1 = this.M1;
        copy.M2 = this.M2;
        copy.M3 = this.M3;
        copy.M4 = this.M4;
        return copy;
    }
}
